Classes/common/touchdb/TD_Database.m

```diff
@@ -193,47 +229,150 @@ - (BOOL)initialize:(NSString*)updates inDatabase:(FMDatabase*)db
+          sqlite3_create_collation(db.sqliteHandle, "JSON", SQLITE_UTF8, kTDCollateJSON_Unicode,
+                                   TDCollateJSON);
+          sqlite3_create_collation(db.sqliteHandle, "JSON_RAW", SQLITE_UTF8, kTDCollateJSON_Raw,
+                                   TDCollateJSON);
+          sqlite3_create_collation(db.sqliteHandle, "JSON_ASCII", SQLITE_UTF8, kTDCollateJSON_ASCII,
+                                   TDCollateJSON);
+          sqlite3_create_collation(db.sqliteHandle, "REVID", SQLITE_UTF8, NULL, TDCollateRevIDs);
+        }];
+    }
 
-    __block BOOL result = YES;
-    __weak TD_Database* weakSelf = self;
-    [_fmdbQueue inDatabase:^(FMDatabase* db) {
-        // Stuff we need to initialize every time the database opens:
-        TD_Database* strongSelf = weakSelf;
-        if (![strongSelf initialize:@"PRAGMA foreign_keys = ON;" inDatabase:db]) result = NO;
-    }];
-    if (!result) {
-        return NO;
+    // Stuff we need to initialize every time the database opens:
+    if (result) {
+        __weak TD_Database* weakSelf = self;
+        [queue inDatabase:^(FMDatabase* db) {
+          TD_Database* strongSelf = weakSelf;
+          if (!strongSelf || ![strongSelf initialize:@"PRAGMA foreign_keys = ON;" inDatabase:db]) {
+              result = NO;
+          }
+        }];
+    }
+    
+    // Assign properties (if everything was OK)
+    if (result) {
+        _fmdbQueue = queue;
+        _keyProviderToOpenDB = provider;
+    } else if (queue) {
+        [queue close];
     }
 
-    return YES;
+    return result;
+}
+
+// callers: many things
+- (BOOL)isOpen
+{
+    return _open;
+}
+
+// callers: many things
+- (BOOL)isOpenWithEncryptionKeyProvider:(id<CDTEncryptionKeyProvider>)provider
+{
+    BOOL isValid = NO;
+    
+    if ([self isOpen]) {
+        isValid = [TD_Database sameEncryptionKeyIn:_keyProviderToOpenDB and:provider];
+        
+        if (!isValid) {
+            CDTLogError(CDTDATASTORE_LOG_CONTEXT, @"DB is open but key provided is wrong");
+        }
+    }
+    
+    return isValid;
+}
+
+// callers: -isOpenWithEncryptionKeyProvider:
++ (BOOL)sameEncryptionKeyIn:(id<CDTEncryptionKeyProvider>)thisProvider
+                        and:(id<CDTEncryptionKeyProvider>)otherProvider
+{
+    NSString *thisKey = [thisProvider encryptionKey];
+    NSString *otherKey = [otherProvider encryptionKey];
+    
+    BOOL sameKey = NO;
+    if (thisKey == nil) {
+        sameKey = (otherKey == nil);
+    } else {
+        sameKey = ((otherKey != nil) && [otherKey isEqualToString:thisKey]);
+    }
+
+    return sameKey;
 }
 
 // callers: many things
-- (BOOL)open
+- (BOOL)openWithEncryptionKeyProvider:(id<CDTEncryptionKeyProvider>)provider
 {
-    if (_open) return YES;
-    if (![self openFMDB]) return NO;
+    Assert(provider, @"Key provider is mandatory. Inform a nil provider instead");
+    
+    if ([self isOpen]) {
+         return [self isOpenWithEncryptionKeyProvider:provider];
+    }
```

Comment:
The intent for method `openWithEncryptionKeyProvider:` is (b).

I think the name already clarifies the purpose because it is not the only method that uses the suffix *WithEncryptionKeyProvider:*, `isOpenWithEncryptionKeyProvider:` also uses it and has the same meaning. Next to this method there is another one without the suffix called `isOpen` that simply verifies if the TD_Database instance is opened with any key.

Also, all these three methods are commented in `TD_Database.h`: [1](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-d0d702218592f847785da5ec2f8d4e99R94), [2](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-d0d702218592f847785da5ec2f8d4e99R83), [3](https://github.com/cloudant/CDTDatastore/pull/112/files#diff-d0d702218592f847785da5ec2f8d4e99R78) and the comments explain how the keyProvider is used.